dolly0920/src/main/java/com/example/dolly0920/week4/Spaceship.java

```diff
@@ -0,0 +1,34 @@
+package com.example.dolly0920.week4;
+
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class Spaceship {
+  private final int id;
+  private final int speed;
+  private final int stability;
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == this) { // previous equals method
+      return true;
+    }
+    if (o == null) { // always return false when parameter is null value
+      return false;
+    }
+    // instanceof : [reference type] instance of [class] : return [boolean]
+    if (!(o instanceof Spaceship)) { // different class => return false
```

Comment:
[difference of instanceof and getclass](http://burucodegallery.blogspot.com/2013/08/4-3-instance-of-getclass.html)

I'm not sure, but it seems a little clearer to use `getClass` when overriding the equals function.
(b/c Instanceof returns true, if classes are in inheritance relationship)
I brought an article about the difference between `getClass` and `instanceof`; I think it would be helpful to read it!

Reply:
Thank you for good comment. I'll check your link and update my code :)

Reply:
I think I gave you the wrong information...😭

https://stackoverflow.com/a/596707/11285469

It is correct to use `getClass()` to check `symmetric`,
BUT When implementing equals, several complex conditions are also taken, and it is said that there is no need to do `getClass` because they checking whether some objects of subclass are the same as objects of superclass..!!

> The reason that I favor the instanceof approach is that when you use the getClass approach, you have the restriction that objects are only equal to other objects of the same class, the same run time type. If you extend a class and add a couple of innocuous methods to it, then check to see whether some object of the subclass is equal to an object of the super class, even if the objects are equal in all important aspects, you will get the surprising answer that they aren't equal. In fact, this violates a strict interpretation of the Liskov substitution principle, and can lead to very surprising behavior. In Java, it's particularly important because most of the collections (HashTable, etc.) are based on the equals method. If you put a member of the super class in a hash table as the key and then look it up using a subclass instance, you won't find it, because they are not equal.
Because of these problems, I didn't even bother discussing the getClass approach. But it turns out that because it does let you add aspects while preserving the equals contract, some people favor it. So I just want to get the information out there that it has disadvantages too. The biggest disadvantage is the fact that you get two objects that appear equal (because they are equal on all the fields) but they are not equal because they are of different classes. This can cause surprising behavior.

Reply:
No... Thanks to your comments, it was nice to be able to think again.